File: ekodb-client-ts/src/query-builder.ts
/**
 * Query Builder for constructing complex queries with fluent API
 */

export enum SortOrder {
  Asc = "asc",
  Desc = "desc",
}

export interface Query {
  filter?: any;
  sort?: Array<{ field: string; ascending: boolean }>;
  limit?: number;
  skip?: number;
  join?: any;
  bypass_cache?: boolean;
  bypass_ripple?: boolean;
}

/**
 * Fluent API for building complex database queries
 *
 * @example
 * ```typescript
 * const query = new QueryBuilder()
 *   .eq("status", "active")
 *   .gt("age", 18)
 *   .sortDesc("created_at")
 *   .limit(10)
 *   .build();
 * ```
 */
export class QueryBuilder {
  private filters: any[] = [];
  private sortFields: Array<{ field: string; order: SortOrder }> = [];
  private _limit?: number;
  private _skip?: number;
  private _join?: any;
  private _bypassCache: boolean = false;
  private _bypassRipple: boolean = false;

  // ========================================================================
  // Comparison Operators
  // ========================================================================

  /**
   * Add an equality filter ($eq)
   */
  eq(field: string, value: any): this {
    this.filters.push({
      type: "Condition",
      content: {
        field,
        operator: "Eq",
        value,
      },
    });
    return this;
  }

  /**
   * Add a not-equal filter ($ne)
   */
  ne(field: string, value: any): this {
    this.filters.push({
      type: "Condition",
      content: {
        field,
        operator: "Ne",
        value,
      },
    });
    return this;
  }

  /**
   * Add a greater-than filter ($gt)
   */
  gt(field: string, value: any): this {
    this.filters.push({
      type: "Condition",
      content: {
        field,
        operator: "Gt",
        value,
      },
    });
    return this;
  }

  /**
   * Add a greater-than-or-equal filter ($gte)
   */
  gte(field: string, value: any): this {
    this.filters.push({
      type: "Condition",
      content: {
        field,
        operator: "Gte",
        value,
      },
    });
    return this;
  }

  /**
   * Add a less-than filter ($lt)
   */
  lt(field: string, value: any): this {
    this.filters.push({
      type: "Condition",
      content: {
        field,
        operator: "Lt",
        value,
      },
    });
    return this;
  }

  /**
   * Add a less-than-or-equal filter ($lte)
   */
  lte(field: string, value: any): this {
    this.filters.push({
      type: "Condition",
      content: {
        field,
        operator: "Lte",
        value,
      },
    });
    return this;
  }

  /**
   * Add an in-array filter ($in)
   */
  in(field: string, values: any[]): this {
    this.filters.push({
      type: "Condition",
      content: {
        field,
        operator: "In",
        value: values,
      },
    });
    return this;
  }

  /**
   * Add a not-in-array filter ($nin)
   */
  nin(field: string, values: any[]): this {
    this.filters.push({
      type: "Condition",
      content: {
        field,
        operator: "NotIn",
        value: values,
      },
    });
    return this;
  }

  // ========================================================================
  // String Operators
  // ========================================================================

  /**
   * Add a contains filter (substring match)
   */
  contains(field: string, substring: string): this {
    this.filters.push({
      type: "Condition",
      content: {
        field,
        operator: "Contains",
        value: substring,
      },
    });
    return this;
  }

  /**
   * Add a starts-with filter
   */
  startsWith(field: string, prefix: string): this {
    this.filters.push({
      type: "Condition",
      content: {
        field,
        operator: "StartsWith",
        value: prefix,
      },
    });
    return this;
  }

  /**
   * Add an ends-with filter
   */
  endsWith(field: string, suffix: string): this {
    this.filters.push({
      type: "Condition",
      content: {
        field,
        operator: "EndsWith",
        value: suffix,
      },
    });
    return this;
  }

  /**
   * Add a regex filter
   */
  regex(field: string, pattern: string): this {
    this.filters.push({
      type: "Condition",
      content: {
        field,
        operator: "Regex",
        value: pattern,
      },
    });
    return this;
  }

  // ========================================================================
  // Logical Operators
  // ========================================================================

  /**
   * Combine filters with AND logic
   */
  and(conditions: any[]): this {
    this.filters.push({
      type: "Logical",
      content: {
        operator: "And",
        expressions: conditions,
      },
    });
    return this;
  }

  /**
   * Combine filters with OR logic
   */
  or(conditions: any[]): this {
    this.filters.push({
      type: "Logical",
      content: {
        operator: "Or",
        expressions: conditions,
      },
    });
    return this;
  }

  /**
   * Negate a filter
   */
  not(condition: any): this {
    this.filters.push({
      type: "Logical",
      content: {
        operator: "Not",
        expressions: [condition],
      },
    });
    return this;
  }

  /**
   * Add a raw filter expression
   */
  rawFilter(filter: any): this {
    this.filters.push(filter);
    return this;
  }

  // ========================================================================
  // Sorting
  // ========================================================================

  /**
   * Add a sort field in ascending order
   */
  sortAsc(field: string): this {
    this.sortFields.push({ field, order: SortOrder.Asc });
    return this;
  }

  /**
   * Add a sort field in descending order
   */
  sortDesc(field: string): this {
    this.sortFields.push({ field, order: SortOrder.Desc });
    return this;
  }

  // ========================================================================
  // Pagination
  // ========================================================================

  /**
   * Set the maximum number of results
   */
  limit(limit: number): this {
    this._limit = limit;
    return this;
  }

  /**
   * Set the number of results to skip (for pagination)
   */
  skip(skip: number): this {
    this._skip = skip;
    return this;
  }

  /**
   * Set page number and page size (convenience method)
   */
  page(page: number, pageSize: number): this {
    this._skip = page * pageSize;
    this._limit = pageSize;
    return this;
  }

  // ========================================================================
  // Joins
  // ========================================================================

  /**
   * Add a join configuration
   */
  join(joinConfig: any): this {
    this._join = joinConfig;
    return this;
  }

  // ========================================================================
  // Performance Flags
  // ========================================================================

  /**
   * Bypass cache for this query
   */
  bypassCache(bypass: boolean = true): this {
    this._bypassCache = bypass;
    return this;
  }

  /**
   * Bypass ripple for this query
   */
  bypassRipple(bypass: boolean = true): this {
    this._bypassRipple = bypass;
    return this;
  }

  // ========================================================================
  // Build
  // ========================================================================

  /**
   * Build the final Query object
   */
  build(): Query {
    const query: Query = {};

    // Combine all filters with AND logic if multiple filters exist
    if (this.filters.length > 0) {
      query.filter =
        this.filters.length === 1
          ? this.filters[0]
          : {
              type: "Logical",
              content: {
                operator: "And",
                expressions: this.filters,
              },
            };
    }

    // Build sort expression
    if (this.sortFields.length > 0) {
      query.sort = this.sortFields.map(({ field, order }) => ({
        field,
        ascending: order === SortOrder.Asc,
      }));
    }

    if (this._limit !== undefined) {
      query.limit = this._limit;
    }

    if (this._skip !== undefined) {
      query.skip = this._skip;
    }

    if (this._join !== undefined) {
      query.join = this._join;
    }

    if (this._bypassCache) {
      query.bypass_cache = true;
    }

    if (this._bypassRipple) {
      query.bypass_ripple = true;
    }

    return query;
  }
}

File: ekodb-client-py/python/ekodb_client/query_builder.py
"""
QueryBuilder for constructing ekoDB queries with a fluent API.

Example:
    ```python
    from ekodb_client import QueryBuilder

    query = (
        QueryBuilder()
        .eq("status", "active")
        .gte("age", 18)
        .sort_desc("created_at")
        .limit(10)
        .build()
    )

    results = await client.find("users", query)
    ```
"""

from typing import Any, Dict, List, Optional


class QueryBuilder:
    """Fluent builder for constructing ekoDB queries."""

    def __init__(self):
        self._filters: List[Dict[str, Any]] = []
        self._sort_fields: List[Dict[str, Any]] = []
        self._limit: Optional[int] = None
        self._skip: Optional[int] = None
        self._join: Optional[Dict[str, Any]] = None
        self._bypass_cache: bool = False
        self._bypass_ripple: bool = False
        self._select_fields: Optional[List[str]] = None
        self._exclude_fields: Optional[List[str]] = None

    # =========================================================================
    # Comparison Operators
    # =========================================================================

    def _add_condition(self, field: str, operator: str, value: Any) -> "QueryBuilder":
        self._filters.append(
            {
                "type": "Condition",
                "content": {
                    "field": field,
                    "operator": operator,
                    "value": value,
                },
            }
        )
        return self

    def eq(self, field: str, value: Any) -> "QueryBuilder":
        """Add an equality filter."""
        return self._add_condition(field, "Eq", value)

    def ne(self, field: str, value: Any) -> "QueryBuilder":
        """Add a not-equal filter."""
        return self._add_condition(field, "Ne", value)

    def gt(self, field: str, value: Any) -> "QueryBuilder":
        """Add a greater-than filter."""
        return self._add_condition(field, "Gt", value)

    def gte(self, field: str, value: Any) -> "QueryBuilder":
        """Add a greater-than-or-equal filter."""
        return self._add_condition(field, "Gte", value)

    def lt(self, field: str, value: Any) -> "QueryBuilder":
        """Add a less-than filter."""
        return self._add_condition(field, "Lt", value)

    def lte(self, field: str, value: Any) -> "QueryBuilder":
        """Add a less-than-or-equal filter."""
        return self._add_condition(field, "Lte", value)

    def in_array(self, field: str, values: List[Any]) -> "QueryBuilder":
        """Add an 'in' filter (field value in list)."""
        return self._add_condition(field, "In", values)

    def nin(self, field: str, values: List[Any]) -> "QueryBuilder":
        """Add a 'not in' filter (field value not in list)."""
        return self._add_condition(field, "NotIn", values)

    def contains(self, field: str, value: str) -> "QueryBuilder":
        """Add a substring contains filter."""
        return self._add_condition(field, "Contains", value)

    def starts_with(self, field: str, value: str) -> "QueryBuilder":
        """Add a starts-with filter."""
        return self._add_condition(field, "StartsWith", value)

    def ends_with(self, field: str, value: str) -> "QueryBuilder":
        """Add an ends-with filter."""
        return self._add_condition(field, "EndsWith", value)

    def regex(self, field: str, pattern: str) -> "QueryBuilder":
        """Add a regex filter (uses Contains operator)."""
        return self._add_condition(field, "Contains", pattern)

    # =========================================================================
    # Logical Operators
    # =========================================================================

    def and_(self, conditions: List[Dict[str, Any]]) -> "QueryBuilder":
        """Combine filters with AND logic."""
        self._filters.append(
            {
                "type": "Logical",
                "content": {
                    "operator": "And",
                    "expressions": conditions,
                },
            }
        )
        return self

    def or_(self, conditions: List[Dict[str, Any]]) -> "QueryBuilder":
        """Combine filters with OR logic."""
        self._filters.append(
            {
                "type": "Logical",
                "content": {
                    "operator": "Or",
                    "expressions": conditions,
                },
            }
        )
        return self

    def not_(self, condition: Dict[str, Any]) -> "QueryBuilder":
        """Negate a filter."""
        self._filters.append(
            {
                "type": "Logical",
                "content": {
                    "operator": "Not",
                    "expressions": [condition],
                },
            }
        )
        return self

    def raw_filter(self, filter_expr: Dict[str, Any]) -> "QueryBuilder":
        """Add a raw filter expression."""
        self._filters.append(filter_expr)
        return self

    # =========================================================================
    # Sorting
    # =========================================================================

    def sort_asc(self, field: str) -> "QueryBuilder":
        """Add ascending sort on a field."""
        self._sort_fields.append({"field": field, "ascending": True})
        return self

    def sort_desc(self, field: str) -> "QueryBuilder":
        """Add descending sort on a field."""
        self._sort_fields.append({"field": field, "ascending": False})
        return self

    # =========================================================================
    # Pagination
    # =========================================================================

    def limit(self, limit: int) -> "QueryBuilder":
        """Set maximum number of results."""
        self._limit = limit
        return self

    def skip(self, skip: int) -> "QueryBuilder":
        """Set number of results to skip."""
        self._skip = skip
        return self

    def page(self, page_num: int, page_size: int) -> "QueryBuilder":
        """Set pagination (page number and size)."""
        self._skip = page_num * page_size
        self._limit = page_size
        return self

    # =========================================================================
    # Joins
    # =========================================================================

    def join(self, join_config: Dict[str, Any]) -> "QueryBuilder":
        """Add a join configuration."""
        self._join = join_config
        return self

    # =========================================================================
    # Performance Flags
    # =========================================================================

    def bypass_cache(self, bypass: bool = True) -> "QueryBuilder":
        """Bypass cache for this query."""
        self._bypass_cache = bypass
        return self

    def bypass_ripple(self, bypass: bool = True) -> "QueryBuilder":
        """Bypass ripple propagation for this query."""
        self._bypass_ripple = bypass
        return self

    # =========================================================================
    # Field Projection
    # =========================================================================

    def select_fields(self, fields: List[str]) -> "QueryBuilder":
        """Select specific fields to return."""
        self._select_fields = fields
        return self

    def exclude_fields(self, fields: List[str]) -> "QueryBuilder":
        """Exclude specific fields from results."""
        self._exclude_fields = fields
        return self

    # =========================================================================
    # Build
    # =========================================================================

    def build(self) -> Dict[str, Any]:
        """Build the final query dict.

        Returns a dict compatible with ekoDB's query format, suitable for
        passing directly to ``client.find(collection, query)``.
        """
        query: Dict[str, Any] = {}

        # Combine filters
        if self._filters:
            if len(self._filters) == 1:
                query["filter"] = self._filters[0]
            else:
                query["filter"] = {
                    "type": "Logical",
                    "content": {
                        "operator": "And",
                        "expressions": self._filters,
                    },
                }

        # Sort
        if self._sort_fields:
            query["sort"] = self._sort_fields

        # Pagination
        if self._limit is not None:
            query["limit"] = self._limit
        if self._skip is not None:
            query["skip"] = self._skip

        # Join
        if self._join is not None:
            query["join"] = self._join

        # Performance flags
        query["bypass_cache"] = self._bypass_cache
        query["bypass_ripple"] = self._bypass_ripple

        # Projection
        if self._select_fields is not None:
            query["select_fields"] = self._select_fields
        if self._exclude_fields is not None:
            query["exclude_fields"] = self._exclude_fields

        return query

    @staticmethod
    def condition(field: str, operator: str, value: Any) -> Dict[str, Any]:
        """Create a standalone condition dict for use with and_/or_/not_.

        Example:
            ```python
            query = (
                QueryBuilder()
                .or_([
                    QueryBuilder.condition("status", "Eq", "active"),
                    QueryBuilder.condition("status", "Eq", "pending"),
                ])
                .build()
            )
            ```
        """
        return {
            "type": "Condition",
            "content": {
                "field": field,
                "operator": operator,
                "value": value,
            },
        }
